messages/off-chain-registry/message_manager.go

```diff
@@ -0,0 +1,198 @@
+package offchainregistry
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	warpPayload "github.com/ava-labs/avalanchego/vms/platformvm/warp/payload"
+	"github.com/ava-labs/awm-relayer/config"
+	"github.com/ava-labs/awm-relayer/vms"
+	"github.com/ava-labs/subnet-evm/accounts/abi/bind"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	teleporterregistry "github.com/ava-labs/teleporter/abi-bindings/go/Teleporter/upgrades/TeleporterRegistry"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+var (
+	OffChainRegistrySourceAddress = common.HexToAddress("0x0000000000000000000000000000000000000000")
+)
+
+const (
+	addProtocolVersionGasLimit  uint64 = 500_000
+	revertVersionNotFoundString        = "TeleporterRegistry: version not found"
+	revertAddressNotFoundString        = "TeleporterRegistry: protocol address not found"
+)
+
+type messageManager struct {
+	logger             logging.Logger
+	destinationClients map[ids.ID]vms.DestinationClient
+	registryAddress    common.Address
+}
+
+func NewMessageManager(
+	logger logging.Logger,
+	messageProtocolConfig config.MessageProtocolConfig,
+	destinationClients map[ids.ID]vms.DestinationClient,
+) (*messageManager, error) {
+	// Marshal the map and unmarshal into the off-chain registry config
+	data, err := json.Marshal(messageProtocolConfig.Settings)
+	if err != nil {
+		logger.Error("Failed to marshal off-chain registry config")
+		return nil, err
+	}
+	var messageConfig Config
+	if err := json.Unmarshal(data, &messageConfig); err != nil {
+		logger.Error("Failed to unmarshal off-chain registry config")
+		return nil, err
+	}
+
+	if err := messageConfig.Validate(); err != nil {
+		logger.Error(
+			"Invalid off-chain registry config.",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	return &messageManager{
+		logger:             logger,
+		destinationClients: destinationClients,
+		registryAddress:    common.HexToAddress(messageConfig.TeleporterRegistryAddress),
+	}, nil
+}
+
+// ShouldSendMessage returns false if the version/protocol address tuple is already registered in the TeleporterRegistry contract.
+func (m *messageManager) ShouldSendMessage(unsignedMessage *warp.UnsignedMessage, destinationBlockchainID ids.ID) (bool, error) {
```

Comment:
Could you add some unit tests for the new message manager implementation (particularly `ShouldSendMessage`)

Reply:
Sure thing, added unit tests in for `ShouldSendMessage`. I omitted them for `SendMessage`, since that's basically a pass through to `DestinationClient.SendTx`, which already has unit tests.